[#39] 결제 기능 구현

wonchaebin opened the pull request:
#39 
결제 기능 구현

- 클라이언트에서 결제 후 imp_uid, merchant_id 부여 받으면 검증 후 주문 상태 변경
- 채팅 결제의 경우 검증 통해 주문이 완료되면 채팅방 id가 부여되어 채팅방 생성
- 결제 방법은 카카오페이와 토스페이먼츠로 나뉘어짐

wonchaebin committed:
Feat: 결제 기본 기능 구현 (#39)

wonchaebin committed:
Merge branch 'dev' of https://github.com/docconneting/docconneting into feat/payment

wonchaebin committed:
Feat: 결제 기능 구현 (#39)

wonchaebin committed:
Fix: 결제 흐름대로 코드 전체적으로 수정 (#39)

wonchaebin committed:
Merge branch 'dev' of https://github.com/docconneting/docconneting into feat/payment

wonchaebin committed:
Feat: 스케줄러 추가로 만료시간 설정 (#39)

suyeonkim56 commented on src/main/java/com/example/docconneting/domain/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,102 @@
+            );
+
+            // 채팅 주문인 경우 채팅방 생성 및 주문에 연결
+            if (order.getOrderType() == OrderType.CHAT) {
```
equals로 통일하자고 했어서, 수정이 필요할 것 같습니다.

wonchaebin replied:
수정했습니다 감사합니다!

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/order/entity/Order.java:
```diff
@@ -11,54 +13,104 @@
 @Entity
 @Table(name = "orders")
 @Getter
 @NoArgsConstructor
```
access level 최소 protected로 해 주세요.

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/order/repository/OrderRepository.java:
```diff
@@ -1,11 +1,44 @@
-}
+    Optional<Order> findByMerchantUid(String merchantUid);
+
+    // 채팅 주문 상태가 COMPLETED이고, chattingRoomId가 아직 비어있는 주문들 중 가장 최근 주문 하나 조회
```
이 쿼리는 1건을 보장하지 않을거 같습니다. 2건 이상이 조회 가능한 쿼리로 보입니다.
jpql을 사용한다면 findLatestCompletedChatOrder 를 list로 받아 첫번째 로우만 리턴하게 하거나
native query를 사용하여 1건이 나오게 해야겠어요.

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/dto/request/PaymentWebhookRequest.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
protected로 access를 제한해 주세요.

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,102 @@
+                paymentMethod,
+                PaymentStatus.COMPLETED,
+                impUid,
+                createdAt
```
AuditingEntityListener 에 approveAt에 createdAt 의 변수를 사용하고 있는것 같습니다. 코드를 잘못인식할 수 있어요.

hyeonseokhoon reviewed (commented):
수고하셨습니다. 코드내 코멘트 확인해 주세요.

wonchaebin committed:
Feat: 키값 처리 로직 추가 및 PortOne 설정 (#39)

suyeonkim56 reviewed (approved):
고생하셨습니다!!

wonchaebin committed:
:wq

11Merge branch 'dev' of https://github.com/docconneting/docconneting into feat/payment

wonchaebin committed:
Feat: 수정된 결제 흐름으로 결제 기능 구현 (#39)

suyeonkim56 reviewed (approved):
고생하셨습니다!!

seiky17 reviewed (approved):
깔끔하게 잘 작성해주셨네요. 고생하셨습니다!

wonchaebin committed:
Merge branch 'dev' of https://github.com/docconneting/docconneting into feat/payment

wonchaebin committed:
Fix: 결제 금액 검증 로직 누락 보완 (#39)

hyeonseokhoon commented on src/main/java/com/example/docconneting/common/config/PortOneProperties.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@Getter
+@ConfigurationProperties(prefix = "portone.api")
```
이건 그냥 알아두시면 좋다는 의미 입니다. 간단하게 골뱅이기호(@)Value 해서 properties를 가져오는 방법도 있습니다.

wonchaebin replied:
헉 알려주셔서 감사합니다!! 😎👏

hyeonseokhoon replied:
밑에 코드까지 보니 특정한 내용을 구분하기 위해 사용했다 볼 수 있겠군요.

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/controller/PaymentController.java:
```diff
@@ -0,0 +1,146 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/payments")
+@Slf4j
```
이건 그냥 보기 좋으라고.. 별것 아닌 내용입니다. 아무런 의미 없기도 하구요. slf4j 를 맨 위로 올리면 피라미드 모양으로 그냥 예뻐 보이는 효과가 있습니다. ㅎ

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/controller/PaymentController.java:
```diff
@@ -0,0 +1,146 @@
+
+    private IamportClient iamportClient;
+
+    @Value("${portone.api.key}")
```
PortOneProperties는 어쩌구요?????

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/controller/PaymentController.java:
```diff
@@ -0,0 +1,146 @@
+                paymentWebhookRequest.getPaymentStatus(),
+                paymentWebhookRequest.getPayMethod());
+        try {
+            Order order = orderRepository.findByMerchantUid(paymentWebhookRequest.getMerchantUid())
```
Controller 에서 Repository 를 DI 하거나 서비스 로직이 들어가는 건 좋아 보이지 않습니다.
Service를 만드시죠

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/payment/controller/PaymentController.java:
```diff
@@ -0,0 +1,146 @@
+    public ResponseEntity<OrderResponse> verifyAndCreateOrder(
+            @RequestBody PaymentVerificationRequest request
+    ) throws IamportResponseException, IOException {
+        Payment payment = portOneService.getPayment(request.getImpUid());
```
여기도 마찬가지 입니다. 컨트롤러가 서비스 로직을 가지는 것은 책임분리에 어울리지 않아요.

wonchaebin replied:
피드백 내용 확인했습니다! 말씀하신 부분 코드 수정하면서 혼용된거 같습니다ㅠㅡㅠ 피드백 감사합니다 튜터님!!!

hyeonseokhoon reviewed (commented):
컨트롤러의 서비스로직, 애써만든 PortOneProperties 와 Value annotation의 혼용 부분을 봐 주세요.

hyeonseokhoon commented:
아.. 수정하셔서 혼용된건가요? ㅋㅋ

wonchaebin committed:
Refactor: 결제 로직 서비스 계층 분리 및 설정 클래스 적용 (#39)

suyeonkim56 reviewed (approved):
고생하셨습니다!!
